.doom.d/config.org
#+TITLE: Doom Emacs Configuration
#+AUTHOR: gty
#+STARTUP: fold

* Step /zero/

Make this file run (slightly) faster with lexical binding (see this [[https://nullprogram.com/blog/2016/12/22/][blog post]]
for more info).

#+begin_src emacs-lisp :comments no
;;; config.el -*- lexical-binding: t; -*-
#+end_src

* Personal information

Some functionality (e.g. =GPG=) uses this to identify me.

#+begin_src emacs-lisp
(setq user-full-name "Tianyu Gu"
      user-mail-address "macdavid313@gmail.com")
#+end_src

* Basics

** Better defaults

These configurations talk for themselves.

#+begin_src emacs-lisp
(setq-default
 delete-by-moving-to-trash t            ; Delete files to trash
 window-combination-resize t ; take new window space from all other windows (not just current)
 x-stretch-cursor t          ; Stretch cursor to the glyph width
 )

(setq undo-limit 80000000               ; Raise undo-limit to 80Mb
      evil-want-fine-undo t ; By default while in insert all changes are one big blob. Be more granular
      auto-save-default t   ; Nobody likes to loose work, I certainly don't
      truncate-string-ellipsis "…" ; Unicode ellispis are nicer than "...", and also save /precious/ space
      )

(display-time-mode 1) ; Enable time in the mode-line

(setq display-time-format "%R%p (%Z)"
      display-time-world-list '(("America/Los_Angeles" "San Francisco")
                                ("Asia/Bangkok" "Bangkok")
                                ("Europe/Kiev" "Kiev")
                                ("Asia/Tokyo" "Tokyo")))

(defun display-time-world-minibuffer ()
  "Like `display-time-world', but just display it in the minibuffer."
  (interactive)
  (message
   (with-temp-buffer
     (display-time-world-display display-time-world-list)
     (buffer-string))))

(global-subword-mode 1) ; Iterate through CamelCase words

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)

;; slightly nicer default buffer names
(setq doom-fallback-buffer-name "► Doom"
      +doom-dashboard-name "► Doom")

(setq confirm-kill-emacs nil) ; ** Don't ask to quit
#+end_src

** Theme

My current desktop is macOS. I use Emacs in the terminal (/iTerm2/) almost
exclusively, and I've found this combination is good enough to me:

+ Desktop background: =Solar Gradients=
+ Terminal theme: =Tomorrow Night Bright=
+ Terminal font: =JetBrainsMono Nerd Font Mono=
+ Emacs theme: =doom-xcode=

#+begin_src emacs-lisp
(setq doom-theme 'doom-xcode)
#+end_src

** Font Face

#+begin_src emacs-lisp
(setq doom-font (font-spec :family "JetBrainsMono Nerd Font Mono" :size 13)
      doom-big-font (font-spec :family "JetBrainsMono Nerd Font Mono" :size 24))
#+end_src

** Open URL in a new buffer

#+begin_src emacs-lisp
(when (fboundp 'eww)
  (defun doom/rename-eww-buffer ()
    "Rename `eww-mode' buffer so sites open in new page.
URL `http://ergoemacs.org/emacs/emacs_eww_web_browser.html'
Version 2017-11-10"
    (let (($title (plist-get eww-data :title)))
      (when (eq major-mode 'eww-mode)
        (if $title
            (rename-buffer (concat "eww " $title ) t)
          (rename-buffer "eww" t)))))

  (add-hook 'eww-after-render-hook 'doom/rename-eww-buffer))
#+end_src

** Helper macros

+ ~load!~ for loading external ~*.el~ files relative to this one
+ ~use-package!~ for configuring packages
+ ~after!~ for running code after a package has loaded
+ ~add-load-path!~ for adding directories to the ~load-path~, relative to this
  file. Emacs searches the ~load-path~ when you load packages with ~require~ or
  ~use-package~.
+ ~map!~ for binding new keys

** Mouse butttons

#+begin_src emacs-lisp
(map! :n [mouse-8] #'better-jumper-jump-backward
      :n [mouse-9] #'better-jumper-jump-forward)
#+end_src

** Window title

#+begin_src emacs-lisp
(setq frame-title-format
      '(""
        (:eval
         (if (s-contains-p org-roam-directory (or buffer-file-name ""))
             (replace-regexp-in-string
              ".*/[0-9]*-?" "☰ "
              (subst-char-in-string ?_ ?  buffer-file-name))
           "%b"))
        (:eval
         (let ((project-name (projectile-project-name)))
           (unless (string= "-" project-name)
             (format (if (buffer-modified-p)  " ◉ %s" "  ●  %s") project-name))))))
#+end_src

** Doom Dashboard

All the useful commands are not really useful to me anymore, so I choose to
disable them.

#+begin_src emacs-lisp
(remove-hook '+doom-dashboard-functions #'doom-dashboard-widget-shortmenu)
(add-hook! '+doom-dashboard-mode-hook (hide-mode-line-mode 1) (hl-line-mode -1))
(setq-hook! '+doom-dashboard-mode-hook evil-normal-state-cursor (list nil))
#+end_src

TODO: customize DOOM dashboard ASCII banner

** Which-key

Let’s make this popup a bit faster

#+begin_src emacs-lisp
(setq which-key-idle-delay 0.5) ;; I need the help, I really do
#+end_src

I also think that having =evil-= appear in so many popups is a bit too verbose,
let’s change that, and do a few other similar tweaks while we’re at it.

#+begin_src emacs-lisp
(setq which-key-allow-multiple-replacements t)
(after! which-key
  (pushnew!
   which-key-replacement-alist
   '(("" . "\\`+?evil[-:]?\\(?:a-\\)?\\(.*\\)") . (nil . "◂\\1"))
   '(("\\`g s" . "\\`evilem--?motion-\\(.*\\)") . (nil . "◃\\1"))
   ))
#+end_src

* Tools

** Abbrev

#+begin_src emacs-lisp
(add-hook 'doom-first-buffer-hook
          (defun +abbrev-file-name ()
            (setq-default abbrev-mode t)
            (setq abbrev-file-name (expand-file-name "abbrev.el" doom-private-dir))))
#+end_src

** Very large files

The very large files mode loads large files in chunks, allowing one to open
ridiculously large files.

To make VLF available without delaying startup, we'll just load it in quiet
moments.

#+begin_src emacs-lisp
(use-package! vlf-setup
  :defer-incrementally vlf-tune vlf-base vlf-write vlf-search vlf-occur vlf-follow vlf-ediff vlf)
#+end_src

** Eros

This package enables the very nice inline evaluation with =gr= and =gR=. The
prefix could be slightly nicer though.

#+begin_src emacs-lisp
(setq eros-eval-result-prefix "⟹ ") ; default =>
#+end_src

** Consult

#+begin_src emacs-lisp
;; (after! consult
;;   (set-face-attribute 'consult-file nil :inherit 'consult-buffer)
;;   (setf (plist-get (alist-get 'perl consult-async-split-styles-alist) :initial) ";"))
#+end_src

** Magit

Magit is absolutely great by default. The diffs don't get any
syntax-highlighting-love though which is a bit sad. Thankfully
[[https://github.com/dandavison/magit-delta][dandavison/magit-delta]] exists, which we can put to use.

** Company

It’s nice to have completions almost all the time, in my opinion. Key strokes
are just waiting to be saved!

#+begin_src emacs-lisp
(after! company
  (setq company-idle-delay 0.5
        company-minimum-prefix-length 2)
  (setq company-show-numbers t)
  (add-hook 'evil-normal-state-entry-hook #'company-abort)) ;; make aborting less annoying.
#+end_src

Now, the improvements from precedent are mostly from remembering history, so
let’s improve that memory.

#+begin_src emacs-lisp
(setq-default history-length 1000)
(setq-default prescient-history-length 1000)
#+end_src

** Plain text

Ispell is nice, let’s have it in text, markdown, and GFM.

#+begin_src emacs-lisp
(set-company-backend!
  '(text-mode
    markdown-mode
    gfm-mode)
  '(:seperate
    company-ispell
    company-files
    company-yasnippet))
#+end_src

** Projectile

#+begin_src emacs-lisp
(setq projectile-indexing-method 'alien)
(setq projectile-sort-order 'recentf)
#+end_src

Looking at documentation via ~SPC h f~ and ~SPC h v~ and looking at the source
can add package src directories to projectile. This isn’t desirable in my
opinion.

#+begin_src emacs-lisp
(setq projectile-ignored-projects '("~/" "/tmp" "~/.emacs.d/" "~/.emacs.d/.local/straight/repos/"))
(defun projectile-ignored-project-function (filepath)
  "Return t if FILEPATH is within any of `projectile-ignored-projects'"
  (or (mapcar (lambda (p) (s-starts-with-p p filepath)) projectile-ignored-projects)))
#+end_src

** String inflection

#+begin_src emacs-lisp
(use-package! string-inflection
  :commands (string-inflection-all-cycle
             string-inflection-toggle
             string-inflection-camelcase
             string-inflection-lower-camelcase
             string-inflection-kebab-case
             string-inflection-underscore
             string-inflection-capital-underscore
             string-inflection-upcase)
  :init
  (map! :leader :prefix ("c~" . "naming convention")
        :desc "cycle" "~" #'string-inflection-all-cycle
        :desc "toggle" "t" #'string-inflection-toggle
        :desc "CamelCase" "c" #'string-inflection-camelcase
        :desc "downCase" "d" #'string-inflection-lower-camelcase
        :desc "kebab-case" "k" #'string-inflection-kebab-case
        :desc "under_score" "_" #'string-inflection-underscore
        :desc "Upper_Score" "u" #'string-inflection-capital-underscore
        :desc "UP_CASE" "U" #'string-inflection-upcase)
  (after! evil
    (evil-define-operator evil-operator-string-inflection (beg end _type)
      "Define a new evil operator that cycles symbol casing."
      :move-point nil
      (interactive "<R>")
      (string-inflection-all-cycle)
      (setq evil-repeat-info '([?g ?~])))
    (define-key evil-normal-state-map (kbd "g~") 'evil-operator-string-inflection)))
#+end_src

** Smart parentheses

#+begin_src emacs-lisp
(sp-local-pair
 '(org-mode)
 "<<" ">>"
 :actions '(insert))
#+end_src

** Tramp

#+begin_quote
TRAMP (Transparent Remote Access, Multiple Protocols) is a package for editing
remote files, similar to AngeFtp or efs. Whereas the others use FTP to connect
to the remote host and to transfer the files, TRAMP uses a remote shell
connection (rlogin, telnet, ssh). It can transfer the files using rcp or a
similar program, or it can encode the file contents (using uuencode or base64)
and transfer them right through the shell connection.
#+end_quote

When using =zsh=, I have to set the value for =tramp-terminal-type= variable ...

#+begin_src emacs-lisp
(setq tramp-terminal-type "tramp")
#+end_src

and also put this into a =~/.zprofile= file at the remote machine.

#+begin_quote
if [ $TERM = tramp ]; then
        unset RPROMPT
        unset RPS1
        PS1="$ "
        unsetopt zle
        unsetopt rcs  # Inhibit loading of further config files
fi
#+end_quote

see [[https://www.emacswiki.org/emacs/TrampMode#h5o-9][here]] for more details.

* Visual

** Emojify

For starters, twitter’s emojis look nicer than emoji-one. Other than that, this
is pretty great OOTB 😀.

#+begin_src emacs-lisp
(setq emojify-emoji-set "twemoji-v2")
#+end_src

One minor annoyance is the use of emojis over the default character when the
default is actually preferred. This occurs with overlay symbols I use in Org
mode, such as checkbox state, and a few other miscellaneous cases.

We can accommodate our preferences by deleting those entries from the emoji hash
table

#+begin_src emacs-lisp
(defvar emojify-disabled-emojis
  '(;; Org
    "◼" "☑" "☸" "⚙" "⏩" "⏪" "⬆" "⬇" "❓"
    ;; Terminal powerline
    "✔"
    ;; Box drawing
    "▶" "◀")
  "Characters that should never be affected by `emojify-mode'.")

(defadvice! emojify-delete-from-data ()
  "Ensure `emojify-disabled-emojis' don't appear in `emojify-emojis'."
  :after #'emojify-set-emoji-data
  (dolist (emoji emojify-disabled-emojis)
    (remhash emoji emojify-emojis)))
#+end_src

** Keycast

For some reason, I find myself demoing Emacs every now and then. Showing what
keyboard stuff I’m doing on-screen seems helpful. While screenkey does exist,
having something that doesn’t cover up screen content is nice.

#+begin_src emacs-lisp
(use-package! keycast
  :commands keycast-mode
  :config
  (define-minor-mode keycast-mode
    "Show current command and its key binding in the mode line."
    :global t
    (if keycast-mode
        (progn
          (add-hook 'pre-command-hook 'keycast--update t)
          (add-to-list 'global-mode-string '("" mode-line-keycast " ")))
      (remove-hook 'pre-command-hook 'keycast--update)
      (setq global-mode-string (remove '("" mode-line-keycast " ") global-mode-string))))
  (custom-set-faces!
    '(keycast-command :inherit doom-modeline-debug
                      :height 0.9)
    '(keycast-key :inherit custom-modified
                  :height 1.1
                  :weight bold)))
#+end_src

** Marginalia

#+begin_src emacs-lisp
(after! marginalia
  (setq marginalia-censor-variables nil)

  (defadvice! +marginalia--anotate-local-file-colorful (cand)
    "Just a more colourful version of `marginalia--anotate-local-file'."
    :override #'marginalia--annotate-local-file
    (when-let (attrs (file-attributes (substitute-in-file-name
                                       (marginalia--full-candidate cand))
                                      'integer))
      (marginalia--fields
       ((marginalia--file-owner attrs)
        :width 12 :face 'marginalia-file-owner)
       ((marginalia--file-modes attrs))
       ((+marginalia-file-size-colorful (file-attribute-size attrs))
        :width 7)
       ((+marginalia--time-colorful (file-attribute-modification-time attrs))
        :width 12))))

  (defun +marginalia--time-colorful (time)
    (let* ((seconds (float-time (time-subtract (current-time) time)))
           (color (doom-blend
                   (face-attribute 'marginalia-date :foreground nil t)
                   (face-attribute 'marginalia-documentation :foreground nil t)
                   (/ 1.0 (log (+ 3 (/ (+ 1 seconds) 345600.0)))))))
      ;; 1 - log(3 + 1/(days + 1)) % grey
      (propertize (marginalia--time time) 'face (list :foreground color))))

  (defun +marginalia-file-size-colorful (size)
    (let* ((size-index (/ (log10 (+ 1 size)) 7.0))
           (color (if (< size-index 10000000) ; 10m
                      (doom-blend 'orange 'green size-index)
                    (doom-blend 'red 'orange (- size-index 1)))))
      (propertize (file-size-human-readable size) 'face (list :foreground color)))))
#+end_src

** All the icons

In some files, =^L= appears as a page break character. This isn’t that visually
appealing, and Steve Purcell has been nice enough to make a package to display
these as horizontal rules.

#+begin_src emacs-lisp
(use-package! page-break-lines
  :commands page-break-lines-mode
  :init
  (autoload 'turn-on-page-break-lines-mode "page-break-lines")
  :config
  (setq page-break-lines-max-width fill-column)
  (map! :prefix "g"
        :desc "Prev page break" :nv "[" #'backward-page
        :desc "Next page break" :nv "]" #'forward-page))
#+end_src

** Treemacs

Quite often there are superfluous files I’m not that interested in. There’s no
good reason for them to take up space. Let’s add a mechanism to ignore them.

#+begin_src emacs-lisp
(after! treemacs
  (defvar treemacs-file-ignore-extensions '()
    "File extension which `treemacs-ignore-filter' will ensure are ignored")
  (defvar treemacs-file-ignore-globs '()

.doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!

;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
(setq user-full-name "Tianyu Gu"
      user-mail-address "macdavid313@gmail.com"
      user-login-name "macdavid313")

;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
;; (setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
;;       doom-variable-pitch-font (font-spec :family "sans" :size 13))

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-xcode)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)

;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

(setq-default
 delete-by-moving-to-trash t                      ; Delete files to trash
 window-combination-resize t                      ; take new window space from all other windows (not just current)
 x-stretch-cursor t)                              ; Stretch cursor to the glyph width

(setq undo-limit 80000000                         ; Raise undo-limit to 80Mb
      evil-want-fine-undo t                       ; By default while in insert all changes are one big blob. Be more granular
      auto-save-default t                         ; Nobody likes to loose work, I certainly don't
      truncate-string-ellipsis "…")               ; Unicode ellispis are nicer than "...", and also save /precious/ space

;; Enable time in the mode-line
(display-time-mode 1)
(setq display-time-format "%R%p (%Z)")

(setq display-time-world-list
      '(("America/Los_Angeles" "San Francisco")
        ("Asia/Bangkok" "Bangkok")
        ("Europe/Kiev" "Kiev")
        ("Asia/Tokyo" "Tokyo")))

(defun display-time-world-minibuffer ()
  "Like `display-time-world', but just display it in the minibuffer."
  (interactive)
  (message
   (with-temp-buffer
     (display-time-world-display display-time-world-list)
     (buffer-string))))

;; Iterate through CamelCase words
(global-subword-mode 1)

;; ** Start maximised/fullscreen (cross-platf)
(add-hook 'window-setup-hook
          (if (eq initial-window-system 'x)
              'toggle-frame-maximized
            'toggle-frame-fullscreen)
          t)

;; open URL in a new buffer
(when (fboundp 'eww)
  (defun doom/rename-eww-buffer ()
    "Rename `eww-mode' buffer so sites open in new page.
URL `http://ergoemacs.org/emacs/emacs_eww_web_browser.html'
Version 2017-11-10"
    (let (($title (plist-get eww-data :title)))
      (when (eq major-mode 'eww-mode)
        (if $title
            (rename-buffer (concat "eww " $title ) t)
          (rename-buffer "eww" t)))))

  (add-hook 'eww-after-render-hook 'doom/rename-eww-buffer))

;; slightly nicer default buffer names
(setq doom-fallback-buffer-name "► Doom"
      +doom-dashboard-name "► Doom")

;; ** Don't ask to quit
(setq confirm-kill-emacs nil)

;;; line number
(setq display-line-numbers-type 'relative)

;;; Projectile
(setq projectile-indexing-method 'alien)
(setq projectile-sort-order 'recentf)

;;; company
(use-package! company
  :config
  (setq company-idle-delay 0.5
        company-minimum-prefix-length 2)
  (setq company-show-numbers t)
  ;; make aborting less annoying.
  (add-hook 'evil-normal-state-entry-hook #'company-abort)
  (setq-default history-length 1000)
  (setq-default prescient-history-length 1000))

;;; LSP
(add-hook 'lsp-mode-hook
          (lambda ()
            (add-to-list 'lsp-file-watch-ignored-directories "[/\\\\]\\.venv\\'")
            (add-to-list 'lsp-file-watch-ignored-directories "[/\\\\]venv\\'")))

;;; Rust
(after! rustic
  (setq rustic-spinner-type 'half-circle
        lsp-rust-analyzer-cargo-watch-command "clippy"
        ;; lsp-eldoc-render-all t
        lsp-rust-analyzer-server-display-inlay-hints t))

;;; Common Lisp and Slime
(add-hook 'lisp-mode-hook #'rainbow-delimiters-mode)

;; for Allegro CL source code
(add-to-list 'auto-mode-alist '("\\.cl\\'" . lisp-mode))

(when (getenv "CL_DOCUMENTATION")
  (let* ((root (file-name-as-directory (getenv "CL_DOCUMENTATION")))
         (acl-doc-root (file-name-as-directory (concat root "allegro")))
         (cltl2-root (file-name-as-directory (concat root "cltl"))))

    (defun acl-doc ()
      "Quickly access AllegroCL Manual"
      (interactive)
      (eww-open-file (concat acl-doc-root "contents.htm")))

    (defun cltl2 ()
      "Quickly access Common Lisp the Language 2nd Edition"
      (interactive)
      (eww-open-file (concat (file-name-as-directory (concat cltl2-root "clm")) "node1.html")))))

(defun sbcl-doc ()
  "Quickly access SBCL Manual"
  (interactive)
  (eww-browse-url "http://www.sbcl.org/manual/index.html"))

(use-package! slime
  :hook (lisp-mode-local-vars . slime-editing-mode)
  :init
  (after! lisp-mode
    (set-repl-handler! 'lisp-mode #'slime-repl)
    (set-eval-handler! 'lisp-mode #'slime-eval-region)
    (set-lookup-handlers! 'lisp-mode
      :definition #'slime-edit-definition
      :documentation #'slime-describe-symbol))
  :config
  (when (getenv "HYPERSPEC_ROOT")
    (let* ((hyperspec-root (file-name-as-directory (getenv "HYPERSPEC_ROOT")))
           (hyperspec-symbol-table (concat hyperspec-root "Data/Map_Sym.txt"))
           (hyperspec-issuex-table (concat hyperspec-root "Data/Map_IssX.txt")))
      (setq common-lisp-hyperspec-root hyperspec-root
            common-lisp-hyperspec-symbol-table hyperspec-symbol-table
            common-lisp-hyperspec-issuex-table hyperspec-issuex-table)))

  (setf slime-repl-history-file (concat doom-cache-dir "slime-repl-history")
        slime-kill-without-query-p t
        slime-default-lisp 'sbcl
        slime-protocol-version 'ignore
        slime-net-coding-system 'utf-8-unix
        slime-load-failed-fasl 'never
        slime-completion-at-point-functions 'slime-fuzzy-complete-symbol)

  (local-set-key [tab] 'slime-complete-symbol)
  (local-set-key (kbd "M-q") 'slime-reindent-defun)

  ;; fix indentation for `if*' when using keyword forms
  (setq common-lisp-indent-if*-keyword
        (rx (? ":") (or "else" "elseif" "then")))

  (set-popup-rules!
    '(("^\\*slime-repl"        :vslot 2 :size 0.3 :quit nil :ttl nil)
      ("^\\*slime-compilation" :vslot 3 :ttl nil)
      ("^\\*slime-traces"      :vslot 4 :ttl nil)
      ("^\\*slime-description" :vslot 5 :size 0.3 :ttl 0)
      ;; Do not display debugger or inspector buffers in a popup window. These
      ;; buffers are meant to be displayed with sufficient vertical space.
      ("^\\*slime-\\(?:db\\|inspector\\)" :ignore t)))

  (let* ((lisp-home (file-name-as-directory (getenv "LISP")))
         (sbcl (concat lisp-home "sbcl/bin/sbcl"))
         (ccl (concat lisp-home "ccl/lx86cl64"))
         (ecl (concat lisp-home "ecl/bin/ecl"))
         (quicklisp-home (concat (file-name-as-directory (getenv "HOME")) "quicklisp"))
         (quicklisp-path (when (file-exists-p quicklisp-home)
                           (concat (file-name-as-directory quicklisp-home)
                                   "setup.lisp"))))

    (when (getenv "ALISP")
      (add-to-list 'slime-lisp-implementations
                   `(alisp ,(remove nil
                                    `(,(getenv "ALISP")
                                      ,@(when quicklisp-path `("-L" ,quicklisp-path)))))))

    (when (getenv "MLISP")
      (add-to-list 'slime-lisp-implementations
                   `(mlisp ,(remove nil
                                    `(,(getenv "MLISP")
                                      ,@(when quicklisp-path `("-L" ,quicklisp-path)))))))

    (when (getenv "ALISP_SMP")
      (add-to-list 'slime-lisp-implementations
                   `(alisp-smp ,(remove nil
                                        `(,(getenv "ALISP_SMP")
                                          ,@(when quicklisp-path `("-L" ,quicklisp-path)))))))

    (when (getenv "MLISP_SMP")
      (add-to-list 'slime-lisp-implementations
                   `(mlisp-smp ,(remove nil
                                        `(,(getenv "MLISP_SMP")
                                          ,@(when quicklisp-path `("-L" ,quicklisp-path)))))))

    (when (file-exists-p sbcl)
      (add-to-list 'slime-lisp-implementations
                   (if quicklisp-path
                       `(sbcl (,sbcl "--load" ,quicklisp-path))
                     `(sbcl (,sbcl "--noinform")))))

    (when (file-exists-p ccl)
      (add-to-list 'slime-lisp-implementations
                   (if quicklisp-path
                       `(ccl (,ccl "--load" ,quicklisp-path))
                     `(ccl (,ccl)))))

    (when (file-exists-p ecl)
      (add-to-list 'slime-lisp-implementations
                   (if quicklisp-path
                       `(ecl (,ecl "-load" ,quicklisp-path))
                     `(ecl (,ecl))))))

  (map! (:map slime-db-mode-map
         :n "gr" #'slime-db-restart-frame)
        (:map slime-inspector-mode-map
         :n "gb" #'slime-inspector-pop
         :n "gr" #'slime-inspector-reinspect
         :n "gR" #'slime-inspector-fetch-all
         :n "K"  #'slime-inspector-describe-inspectee)
        (:map slime-xref-mode-map
         :n "gr" #'slime-recompile-xref
         :n "gR" #'slime-recompile-all-xrefs)
        (:map lisp-mode-map
         :n "gb" #'slime-pop-find-definition-stack)

        (:localleader
         :map lisp-mode-map
         :desc "slime"          "'" #'slime
         :desc "slime (ask)"    ";" (cmd!! #'slime '-)
         :desc "Expand macro" "m" #'macrostep-expand
         (:prefix ("c" . "compile")
          :desc "Compile file"          "c" #'slime-compile-file
          :desc "Compile/load file"     "C" #'slime-compile-and-load-file
          :desc "Compile toplevel form" "f" #'slime-compile-defun
          :desc "Load file"             "l" #'slime-load-file
          :desc "Remove notes"          "n" #'slime-remove-notes
          :desc "Compile region"        "r" #'slime-compile-region)
         (:prefix ("e" . "evaluate")
          :desc "Evaluate buffer"     "b" #'slime-eval-buffer
          :desc "Evaluate last"       "e" #'slime-eval-last-expression
          :desc "Evaluate/print last" "E" #'slime-eval-print-last-expression
          :desc "Evaluate defun"      "f" #'slime-eval-defun
          :desc "Undefine function"   "F" #'slime-undefine-function
          :desc "Evaluate region"     "r" #'slime-eval-region)
         (:prefix ("g" . "goto")
          :desc "Go back"              "b" #'slime-pop-find-definition-stack
          :desc "Go to"                "d" #'slime-edit-definition
          :desc "Go to (other window)" "D" #'slime-edit-definition-other-window
          :desc "Next note"            "n" #'slime-next-note
          :desc "Previous note"        "N" #'slime-previous-note
          :desc "Next sticker"         "s" #'slime-stickers-next-sticker
          :desc "Previous sticker"     "S" #'slime-stickers-prev-sticker)
         (:prefix ("h" . "help")
          :desc "Who calls"               "<" #'slime-who-calls
          :desc "Calls who"               ">" #'slime-calls-who
          :desc "Lookup format directive" "~" #'hyperspec-lookup-format
          :desc "Lookup reader macro"     "#" #'hyperspec-lookup-reader-macro
          :desc "Apropos"                 "a" #'slime-apropos
          :desc "Who binds"               "b" #'slime-who-binds
          :desc "Disassemble symbol"      "d" #'slime-disassemble-symbol
          :desc "Describe symbol"         "h" #'slime-describe-symbol
          :desc "HyperSpec lookup"        "H" #'slime-hyperspec-lookup
          :desc "Who macro-expands"       "m" #'slime-who-macroexpands
          :desc "Apropos package"         "p" #'slime-apropos-package
          :desc "Who references"          "r" #'slime-who-references
          :desc "Who specializes"         "s" #'slime-who-specializes
          :desc "Who sets"                "S" #'slime-who-sets)
         (:prefix ("r" . "repl")
          :desc "Clear REPL"         "c" #'slime-repl-clear-repl
          :desc "Quit connection"    "q" #'slime-quit-lisp
          :desc "Restart connection" "r" #'slime-restart-inferior-lisp
          :desc "Sync REPL"          "s" #'slime-repl-sync)
         (:prefix ("s" . "stickers")
          :desc "Toggle breaking stickers" "b" #'slime-stickers-toggle-break-on-stickers
          :desc "Clear defun stickers"     "c" #'slime-stickers-clear-defun-stickers
          :desc "Clear buffer stickers"    "C" #'slime-stickers-clear-buffer-stickers
          :desc "Fetch stickers"           "f" #'slime-stickers-fetch
          :desc "Replay stickers"          "r" #'slime-stickers-replay
          :desc "Add/remove sticker"       "s" #'slime-stickers-dwim)
         (:prefix ("t" . "trace")
          :desc "Toggle"         "t" #'slime-toggle-trace-fdefinition
          :desc "Toggle (fancy)" "T" #'slime-toggle-fancy-trace
          :desc "Untrace all"    "u" #'slime-untrace-all)))

  (when (featurep! :editor evil +everywhere)
    (add-hook 'slime-mode-hook #'evil-normalize-keymaps)))

(use-package! slime-company
  :after (slime company)
  :init (slime-setup '(slime-fancy slime-company slime-fuzzy
                                   slime-asdf slime-banner slime-xref-browser slime-scratch
                                   slime-trace-dialog slime-mdot-fu))
  :config
  (setq slime-company-completion 'fuzzy
        slime-company-after-completion 'slime-company-just-one-space)
  (define-key company-active-map (kbd "\C-n") 'company-select-next)
  (define-key company-active-map (kbd "\C-p") 'company-select-previous)
  (define-key company-active-map (kbd "\C-d") 'company-show-doc-buffer)
  (define-key company-active-map (kbd "M-.") 'company-show-location))

;;;; Racket
;; (add-to-list 'auto-mode-alist '("\\.rkt\\'" . racket-mode))

(after! racket-mode
  (add-hook! racket-mode
             #'racket-smart-open-bracket-mode))

(add-hook 'racket-mode-hook      #'racket-unicode-input-method-enable)
(add-hook 'racket-repl-mode-hook #'racket-unicode-input-method-enable)

;;;; Paredit
(use-package! paredit
  :config
  ;; hook lisp modes
  (add-hook 'emacs-lisp-mode-hook       #'enable-paredit-mode)
  (add-hook 'eval-expression-minibuffer-setup-hook #'enable-paredit-mode)
  (add-hook 'ielm-mode-hook             #'enable-paredit-mode)
  (add-hook 'lisp-mode-hook             #'enable-paredit-mode)
  (add-hook 'lisp-interaction-mode-hook #'enable-paredit-mode)
  (add-hook 'racket-mode-hook           #'enable-paredit-mode)
  ;; Electric RETURN
  (defvar electrify-return-match
    "[\]}\)\"]"
    "If this regexp matches the text after the cursor, do an \"electric\"
  return.")
  (defun electrify-return-if-match (arg)
    "If the text after the cursor matches `electrify-return-match' then
  open and indent an empty line between the cursor and the text.  Move the
  cursor to the new line."
    (interactive "P")
    (let ((case-fold-search nil))
      (if (looking-at electrify-return-match)
      (save-excursion (newline-and-indent)))
      (newline arg)
      (indent-according-to-mode)))
  ;; Using local-set-key in a mode-hook is a better idea.
  (global-set-key (kbd "RET") 'electrify-return-if-match))

;;;; SPARQL-MODE
(use-package! sparql-mode
  :config
  (add-hook 'auto-mode-alist '("\\.sparql$" . sparql-mode))
  (add-to-list 'auto-mode-alist '("\\.rq$" . sparql-mode))
  (add-hook 'sparql-mode-hook 'company-mode))

;;;; org-roam v2
(setq org-roam-directory "~/org/roam")

;;;; org-journal
(setq org-journal-dir "~/org/journal"
      org-journal-date-prefix "#+TITLE: "
      org-journal-time-prefix "* "
      org-journal-file-format "%Y-%m-%d.org")

;;;; elfeed
(setq rmh-elfeed-org-files '("~/.config/org-elfeed.org"))
(add-hook! 'elfeed-search-mode-hook 'elfeed-update)

;;;; org-mode Agenda
(setq org-agenda-files '("~/org/agenda/"))

;;;; calendar
(setq calendar-month-name-array         ; Month
      ["January" "February" "March"     "April"   "May"      "June"
       "July"    "August"   "September" "October" "November" "December"]

      calendar-day-name-array           ; Week days
      ["Sunday" "Monday" "Tuesday" "Wednesday" "Thursday" "Friday" "Saturday"]

      calendar-week-start-day        ; First day of the week, 0:Sunday, 1:Monday
      1)

(defun open-calendar ()
  (interactive)
  (cfw:open-calendar-buffer
   :contents-sources
   (list
    (cfw:org-create-source "Orange")
    (cfw:ical-create-source "中国节假日" "https://calendar.google.com/calendar/ical/zh-cn.china%23holiday%40group.v.calendar.google.com/public/basic.ics" "Red")
    (cfw:ical-create-source "Holidays in Australia" "https://calendar.google.com/calendar/ical/zh-cn.australian%23holiday%40group.v.calendar.google.com/public/basic.ics" "Blue"))))
